### PR TITLE
Optimize UFS access on missing paths

### PR DESCRIPTION
**Pull request by ZacBlanco**
The main contribution of this PR is to remove a 2nd UFS call against the UFS when performing a listStatus/getStatus against a path which doesn't exist in Alluxio or the UFS.  Previously, `fetchStatusIfAbsent` would return null, and then call `loadMetadata` with a null status which then would make an `exists` call against the UFS. Now we will instead throw a FileNotFoundException which is the same behavior as specified in the UFS API.

### CODE REVIEW

**Comment by ZacBlanco** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java [hunk 1]
```diff
@@ -152,15 +153,15 @@ public UfsStatus fetchStatusIfAbsent(AlluxioURI path, MountTable mountTable)
...
+        throw new FileNotFoundException("fetched status is null for: " + ufsUri);
       }
       ufsStatus.setName(path.getName());
       addStatus(path, ufsStatus);
       return ufsStatus;
     } catch (IllegalArgumentException | IOException e) {
       LogUtils.warnWithException(LOG, "Failed to fetch status for {}", path, e);
     }
-    return null;
+    throw new FileNotFoundException("Failed to fetch status of " + ufsUri);
```
Comment (ggezer) [inline core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:164]: is this retried on caller site?

File: core/server/master/src/test/java/alluxio/underfs/UfsStatusCacheTest.java [hunk 2]
```diff
@@ -371,18 +372,26 @@ public void testFetchExistingSingleStatus() throws Exception {
   @Test
   public void testFetchSingleStatusNonExistingPath() throws Exception {
     spyUfs();
-    UfsStatus fetched = mCache.fetchStatusIfAbsent(new AlluxioURI("/testFile"), mMountTable);
+    try {
+      UfsStatus fetched = mCache.fetchStatusIfAbsent(new AlluxioURI("/testFile"), mMountTable);
+      fail("Should have thrown FileNotFoundException");
```
Comment (ggezer) [inline core/server/master/src/test/java/alluxio/underfs/UfsStatusCacheTest.java:377]: Can you replace these with`ExpectedException` rule?

File: core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java (hunk 1 above)
Comment (ZacBlanco) [reply core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:164]: no

File: core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java [hunk 3]
```diff
@@ -152,15 +153,15 @@ public UfsStatus fetchStatusIfAbsent(AlluxioURI path, MountTable mountTable)
       UnderFileSystem ufs = ufsResource.get();
       UfsStatus ufsStatus = ufs.getStatus(ufsUri.toString());
       if (ufsStatus == null) {
-        return null;
+        throw new FileNotFoundException("fetched status is null for: " + ufsUri);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:156]: Will this exception be thrown to the client? If so we should make it more explanatory (ie. what does `null` status mean?)

Comment (ZacBlanco) [reply core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:156]: No, this is not returned to the client

File: core/server/master/src/test/java/alluxio/underfs/UfsStatusCacheTest.java (hunk 2 above)
Comment (ZacBlanco) [reply core/server/master/src/test/java/alluxio/underfs/UfsStatusCacheTest.java:377]: Using the ExpectedException rule I can't use `Mockito.verify` afterwards

**Comment by gpang** [comment]
@ZacBlanco This still has the commits of https://github.com/Alluxio/alluxio/pull/11720 ?

**Comment by ZacBlanco** [comment]
Sorry, forgot to push

**Comment by ZacBlanco** [comment]
alluxio-bot, test this please

**Comment by ZacBlanco** [comment]
jenkins, test this please

**Comment by ZacBlanco** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java (hunk 3 above)
Comment (calvinjia) [reply core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:156]: Just want to double check, I don't see `FileNotFoundException` being caught by the callers? ie. `processSyncPath`

Comment (ZacBlanco) [reply core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java:156]: `FileNotFoundException` extends from `IOException` which is in the method definition of the caller `loadMetadataForPath` as a part of it's `throws` declarations.

`IOException` is then caught by the try-catch block on line 422 in `processSyncPath` or if the root path is missing, then it is caught in the handler on line 300

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.

**Comment by calvinjia** [comment]
alluxio-bot, cherry-pick this to branch-2.3 please.

**Pull request merged**